### ISSUE TITLE
[RFE] Add states for non-blocking actions in the API

### ISSUE THREAD

**Issue by dyasny** [scylladb/scylla#7742]
We have several actions that are non-blocking (example: refresh) 

I would like to be able to poll their status (progress would be even better but harder) from the API so that I don't run the same action twice. 

In other words, as a user I want to be able to see if a refresh, cleanup, etc are actually running in a given context (node/ks/table).

**Comment by dyasny** [comment]
@vladzcloudius FYI opened as per request

**Comment by vladzcloudius** [comment]
@slivne @amnonh FYI

**Comment by tzach** [comment]
@bhalevy I think this is very similar to the Task Manager

**Comment by bhalevy** [comment]
> @bhalevy I think this is very similar to the Task Manager

Right, indeed.  The plan is to track non-blocking operations, like RBNO and the likes with the task manager (https://github.com/scylladb/scylladb/issues/9809).

That said, the subject of this issue is too broad.

@Deexie is currently working on integrating the task manager with repair tasks, and that will include all repair-based node operations on top of user repairs that are initiated over the api.

The next step is integrating the task manager with compaction to track user-initiated (major, cleanup, upgrade) compaction tasks as well as internally-initiated (off-strategy and regular) compaction tasks.

@vladzcloudius, what else is required? in in what priority?

**Comment by vladzcloudius** [comment]
> what else is required? in in what priority?

@bhalevy 
We need user initiated compactions become non-blocking a year (or more?) ago - this is blocking a SM major compaction integration and makes us do weird things in automations related to cleanups for example.

This is one of the initial "Top 5 issues"!

This is the highest priority and I'm surprised that you are still asking about this.
And clearly the priority of this doesn't correspond to its importance.

@dorlaor @harel-z @DoronArazii FYI

**Comment by bhalevy** [comment]
I'll take the above as "no" to anything else missing.
If anything comes into mind, please open a particular issue for it

**Comment by vladzcloudius** [comment]
> If anything comes into mind, please open a particular issue for it

To make it very simple: every REST API should be asynchronous, not just repairs and compactions, e.g. decommission, removenode, etc.

> I'll take the above as "no" to anything else missing.

I was referring the "in what priority?" part of your original question in my previous post, @bhalevy.

A lot is missing - the problem is that almost 2 years have passed since this issue was opened and all we've got since then is only a repair status.

Please, see the description of this issue - it talks about APIs *in general*.
So, every REST API that doesn't return an immediate result should be asynchronous and have a corresponding REST API to check the status of the corresponding async task.

Do you really want me to start opening GH issues for every such REST API or you can handle it on your own since you seems to be working on a generic framework?

**Comment by vladzcloudius** [comment]
> Modifying the whole API as you suggested is unrealistic, as this will require changing jmx to satisfy the synchronous nodetool semantics.

You don't have to change `nodetool`, `jmx` or the APIs they use - this is NOT we are asking for here.
We need APIs (they may have a different name, e.g. `<current_api_name>_async`) that would satisfy requirements as I have described above.

I don't see it as unrealistic.

> Note that with the task manager, even synchronous api calls may be tracked asynchronously and aborted in some cases, whild they are executing, without the need to make them asynchronous.

This is not going to help us much. Canceling the task is not the main use case this issue is about.
The APIs that we use in automations MUST become asynchronous - nothing less than that is going to be satisfactory.

The bottom line: my original point stands. Let me rephrase it a little to avoid any confusion on your end: we need all currently synchronous APIs that schedule tasks today to have a corresponding asynchronous API as described above.

> A catch-all issue like this one is impossible to track and plan around.

I can see your point.
Please, don't stop yourself from creating a separate GH issue for every single API if/when you decide to work on or when you create a working plan.

We need all of them at the end - hence this issue.
